src/models/model.cpp

```diff
@@ -249,32 +249,159 @@ int32_t Tokenizer::TokenToTokenId(const char* token) const {
+static const uint8_t g_trivial_model[] = {
+    0x08, 0x0a, 0x12, 0x01, 0x61, 0x3a, 0x53, 0x0a, 0x38, 0x12, 0x06, 0x76, 0x61, 0x6c, 0x75, 0x65,
+    0x73, 0x22, 0x08, 0x43, 0x6f, 0x6e, 0x73, 0x74, 0x61, 0x6e, 0x74, 0x2a, 0x24, 0x0a, 0x05, 0x76,
+    0x61, 0x6c, 0x75, 0x65, 0x2a, 0x18, 0x08, 0x01, 0x10, 0x01, 0x42, 0x0c, 0x63, 0x6f, 0x6e, 0x73,
+    0x74, 0x5f, 0x74, 0x65, 0x6e, 0x73, 0x6f, 0x72, 0x4a, 0x04, 0x00, 0x00, 0x00, 0x00, 0xa0, 0x01,
+    0x04, 0x12, 0x01, 0x62, 0x62, 0x14, 0x0a, 0x06, 0x76, 0x61, 0x6c, 0x75, 0x65, 0x73, 0x12, 0x0a,
+    0x0a, 0x08, 0x08, 0x01, 0x12, 0x04, 0x0a, 0x02, 0x08, 0x01, 0x42, 0x04, 0x0a, 0x00, 0x10, 0x15};
+
 // Since Python/Others can and will hold onto a generator object past the model object's lifetime we need to ensure
 // the allocator used is not destroyed until last. This keeps the allocator around until exit, after all other memory
 // has been destroyed. Without this, we will crash in the Onnxruntime BFCArena code when deleting tensors due to the
 // arena already being destroyed.
-void EnsureDeviceOrtInit(OrtSession& session, DeviceType type) {
+void EnsureDeviceOrtInit(DeviceType type) {
   // CPU Allocator is a special case, it's not in the owned 'allocator_device_' table below so we handle it separately
   if (type == DeviceType::CPU)
     return;
 
-  auto& device = GetOrtGlobals()->allocator_device_[static_cast<int>(type)];
-  if (device)
+  auto& allocator = GetOrtGlobals()->device_allocators_[static_cast<int>(type)];
+  if (allocator.allocator_)
     return;
 
+  auto session_options = OrtSessionOptions::Create();
+
+  const char* device_name{};
+  switch (type) {
+    case DeviceType::CUDA:
+      device_name = "cuda";
+      break;
+    case DeviceType::DML:
+      device_name = "dml";
+      break;
+    case DeviceType::WEBGPU:
+      device_name = "webgpu";
+      break;
+    case DeviceType::QNN:
+      device_name = "qnn";
+      break;
+    default:
+      throw std::runtime_error("Unknown device type");
+  }
+
+  std::vector<Config::ProviderOptions> provider_options_list;
+  provider_options_list.emplace_back(Config::ProviderOptions{device_name, {}});
+
+  SetProviderSessionOptions(*session_options, provider_options_list, true, false);
+  allocator.session_ = OrtSession::Create(GetOrtEnv(), g_trivial_model, sizeof(g_trivial_model), session_options.get());
   static const char* device_type_names[] = {"CPU (Not used, see above)", "Cuda", "DML", "WebGPU_Buffer", "QnnHtpShared"};
   static_assert(std::size(device_type_names) == static_cast<size_t>(DeviceType::MAX));
 
   auto name = device_type_names[static_cast<int>(type)];
   auto memory_info = OrtMemoryInfo::Create(name, OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemType::OrtMemTypeDefault);
-  device = Ort::Allocator::Create(session, *memory_info);
-  if (!device)
+  allocator.allocator_ = Ort::Allocator::Create(*allocator.session_, *memory_info);
+  if (!allocator.allocator_)
```

Comment:
if this is going to throw should we reset allocator.session_?

Reply:
It creates the OrtSession a few lines earlier. If someone resets the session_ the allocator is now invalid. Why would that happen?

Reply:
If the allocator was not successfully created and this code is going to throw, do we want allocator.session_ to remain? I would assume it's bad for some reason. It will get replaced on the next call to EnsureDeviceOrtInit but why wait?

Reply:
Changed to reset everything on failure.